File: presentation/rockets/population.js
class Population {
  constructor() {
    this.pop = [];
  }

  /*
   * Initialize population, create new rockets
   */
  init() {
    for (let i = 0; i < populationSize; i += 1) {
      let r = new Rocket();
      r.initDNA();
      this.pop.push(r);
    }
  }

  /*
   * updat the population, draw the rockets, check amount alive
   */
  update() {
    this.alive_count = 0;
    for (let i = 0; i < this.pop.length; i += 1) {
      this.pop[i].update();
      this.pop[i].check();
      this.pop[i].draw();
      if (this.pop[i].alive) {
        this.alive_count += 1;
      }
    }
  }

  /*
   * Splice DNA, per gene random from one of the parents
   * Then apply mutation based on mutation rate
   */
  offspring(a, b) {
    let newGenome = [];
    for (let i = 0; i < a.dna.length; i++) {
      let r = random();
      if (r <= 0.5) {
        newGenome.push(a.dna[i]);
      } else {
        newGenome.push(b.dna[i]);
      }
    }
    for (let i = 0; i < newGenome.length; i++) {
      if (random <= mutationRate) {
        newGenome[i] = p5.Vector.random2D();
      }
    }
    return newGenome;
  }

  /*
   * Create new dna from two parents
   */
  newPopulation() {
    this.createPool();
    let p = [];
    for (let i = 0; i < populationSize; i++) {
      let a = this.pool[floor(random(this.pool.length))];
      let b = this.pool[floor(random(this.pool.length))];
      let dna = this.offspring(a, b);
      let r = new Rocket();
      r.dna = dna;
      p.push(r);
    }
    this.pop = p;
  }

  /*
   * Create a pool based on fitness values
   */
  createPool() {
    this.pool = [];
    for (let i = 0; i < this.pop.length; i++) {
      let n = Math.floor(this.pop[i].fitness * 100);
      for (let j = 0; j < n; j++) {
        this.pool.push(this.pop[i]);
      }
    }
  }
}

File: code/to_be/population.py
"""population
population class for genetic algorithm. Contains initiation functions,
calls for fitness fvalues, creates genepool and regulates reproduction.
"""

from typing import List
import numpy as np

from individual import Individual


class Population():
    """
    Control population during running the genetic algorithm. Has 
    functions for creating a population, producing offspring and
    scoring the population.

    Parameters:
        genes (str): options to create dna from
        population_size (int): amount of individuals per generation
        target (str): target of evolution
        target_length (int): length of target
        mutation_rate (float): determines mutation during reproduction
    """

    def __init__(self,
                 genes: str,
                 population_size: int,
                 target: str,
                 target_length: int,
                 mutation_rate: float) -> None:

        self.genes = genes

        self.population: List[Individual] = []
        self.population_size: int = population_size

        self.target: str = target
        self.target_length: int = target_length

        self.mutation_rate: float = mutation_rate

        self.max_fitness: float = 0.0

    def init_population(self) -> None:
        """
        Create n individuals where n is the population
        size. Save these in self.population.
        """

        for _ in range(self.population_size):
            self.population.append(Individual(
                self.genes,
                self.target,
                self.target_length))


    def _create_p_list(self) -> None:
        """
        Create a list of fitness scores and scale it to have a sum
        of 1.
        """

        p_list = np.array([ind.fitness for ind in self.population])
        return p_list / np.sum(p_list)


    def create_offspring(self) -> None:
        """
        For population size, pick two individuals from the populatoin
        based on the fitness score and create offspring, this will be
        the new population.
        """

        new_population = []
        p_list = self._create_p_list()
        for _ in range(self.population_size):
            # Select to individuals, chance is scaled to fitness score
            parents = np.random.choice(self.population, 2, p=p_list)
            # Create new dna form parents
            new_dna = parents[0].mix_dna(
                parents[1].dna, self.mutation_rate)
            new_population.append(Individual(
                self.genes,
                self.target,
                self.target_length,
                new_dna))
        self.population = new_population


    def calc_fitness(self) -> None:
        """
        Sort the population and extract the maximal fitness from
        the population
        """
        self.population = sorted(
            self.population,
            key=lambda x: x.fitness,
            reverse=True)
        self.max_fitness = self.population[0].fitness
